corehq/apps/reports/tests/test_case_data.py

```diff
@@ -487,3 +487,17 @@ def test_case_deletion_redirect_if_case_is_already_deleted(self):
 
         return_dict = view.get_cases_and_forms_for_deletion(request, self.domain, case.case_id)
         self.assertTrue(return_dict['redirect'])
+
+    def test_forms_unarchived_and_nothing_deleted_if_archive_fails_midway(self):
+        request = deepcopy(self.request)
+        setattr(request, 'session', {'samlSessionIndex': None})
+        setattr(request, '_messages', FallbackStorage(request))
+
+        cases, xforms = self.make_complex_case()
+        _, error = soft_delete_cases_and_forms(request, self.domain,
+                                               list(cases.values()), list(xforms.values())[:-1])
```

Comment:
in making the multiple cases/forms in `make_complex_case`, the last form updates a particular child case while the second form creates it. Since on the first error it'll sort it in reverse `received_on` order, it'll at least archive the the third form before throwing an error saying it can't archive the second form, since the fourth form wasn't archived/part of the delete_form_list.